q2_annotate/semibin2/semibin2.py

```diff
@@ -0,0 +1,271 @@
+import tempfile
+from copy import deepcopy
+
+import skbio.io
+from q2_types.feature_data import DNAIterator
+
+from q2_types.per_sample_sequences import ContigSequencesDirFmt, BAMDirFmt
+from q2_types.per_sample_sequences import MultiFASTADirectoryFormat
+
+from q2_annotate._utils import run_command, _process_common_input_params
+from q2_annotate.semibin2.utils import _process_semibin2_arg
+
+
+def _assert_samples(contigs, maps) -> dict:
+    """Validate that contigs and alignment maps belong to the same sample set."""
+    contig_fps = contigs.sample_dict().values()
+    map_fps = glob.glob(os.path.join(str(maps), "*.bam"))
+    contig_fps, map_fps = sorted(contig_fps), sorted(map_fps)
+
+    contig_samples = contigs.sample_dict().keys()
+    map_samples = [Path(fp).stem.rsplit("_alignment", 1)[0] for fp in map_fps]
+    if set(contig_samples) != set(map_samples):
+        raise Exception(
+            "Contigs and alignment maps should belong to the same sample set. "
+            f'You provided contigs for samples: {",".join(contig_samples)} '
+            f'but maps for samples: {",".join(map_samples)}. Please check '
+            "your inputs and try again."
+        )
+
+    return {
+        s: {"contigs": contig_fps[i], "map": map_fps[i]}
+        for i, s in enumerate(contig_samples)
+    }
+
+
+def _sort_bams(samp_name, samp_props, loc):
+    """Sort BAM files for SemiBin2 processing."""
+    sorted_bam = os.path.join(loc, f"{samp_name}_alignment_sorted.bam")
+    new_props = deepcopy(samp_props)
+    run_command(["samtools", "sort", new_props["map"], "-o", sorted_bam], verbose=True)
+    new_props["map"] = sorted_bam
+    return new_props
+
+
+def _run_semibin2_single(samp_name, samp_props, loc, common_args, multi_sample=False):
+    """Run SemiBin2 in single-sample mode."""
+    bins_dp = os.path.join(loc, samp_name)
+    os.makedirs(bins_dp, exist_ok=True)
+    
+    cmd = [
+        "SemiBin2",
+        "single_easy_bin" if not multi_sample else "multi_easy_bin",
+        "-i", samp_props["contigs"],
+        "-b", samp_props["map"],
+        "-o", bins_dp,
+        "--verbose",  # Always run in verbose mode as specified in requirements
+    ]
+    cmd.extend(common_args)
+    run_command(cmd, verbose=True)
+    return bins_dp
+
+
+def _concatenate_contigs_with_semibin2(sample_set, loc):
+    """Concatenate contigs using SemiBin2's concatenate_fasta command."""
+    # Collect all contig file paths
+    contig_files = []
+    for samp_name, props in sample_set.items():
+        if os.path.exists(props["contigs"]):
+            contig_files.append(props["contigs"])
+    
+    # Use SemiBin2 concatenate_fasta to combine and rename contigs
+    # The -o parameter should be a directory, and concatenated.fa will be created inside
+    concatenate_dir = os.path.join(loc, "concatenated_contigs")
+    os.makedirs(concatenate_dir, exist_ok=True)
+    cmd = ["SemiBin2", "concatenate_fasta", "-i"] + contig_files + ["-o", concatenate_dir]
```

Comment:
Can you adjust this so that we do the concatenation not using the concatenate_fasta script? Do not remove this code, though - just comment it out and implement our own concatenation. We have this method available in q2_assembly._utils which you could perhaps use?

```
def concatenate_files(input_files, output_file):
    """Concatenate the content of the files in input_files and
    save the content in the output_file.

    Args:
        input_files (list): list of all files to be concatenated
        output_file (str): the path to the resulting file
    """
    cmd = ["cat", *input_files]
    subprocess.run(cmd, stdout=open(output_file, "w"), check=True)
```